thesis-gui/src/components/private-quiz/PrivateQuiz.tsx

```diff
@@ -0,0 +1,178 @@
+const NEW_QUESTION_DELAY_MS = 1500;
+
+const WrongKeyErrorCard: React.FC<{ secretKey: string }> = ({ secretKey }) => {
+    return (
+        <IonCard>
+            <IonCardHeader>
+                <IonCardSubtitle>Error</IonCardSubtitle>
+                <IonCardTitle>Wrong Key</IonCardTitle>
+            </IonCardHeader>
+
+            <IonCardContent>
+                { `There's no private quiz with secret key '${secretKey}'!\nGet the right one and try again ` }
+                <IonRouterLink routerLink="/private-quiz">here</IonRouterLink>
+            </IonCardContent>
+        </IonCard>
+    );
+}
+
+const QuizResultCard: React.FC<{
+    questionsSetName: string,
+    correctAnswers: number,
+    numberOfQuestions: number
+}> = ({
+        questionsSetName,
+        correctAnswers,
+        numberOfQuestions
+}) => {
+    return (
+        <IonCard>
+            <IonCardHeader>
+                <IonCardSubtitle>
+                    { `Quiz: ${ questionsSetName }` }
+                </IonCardSubtitle>
+                <IonCardTitle>
+                    You finished the quiz!
+                </IonCardTitle>
+            </IonCardHeader>
+            <IonCardContent>
+                { `Your result is: ${correctAnswers} / ${numberOfQuestions} (${Math.round(correctAnswers / numberOfQuestions * 100)}%)` }
+            </IonCardContent>
+        </IonCard>
+    );
+}
+
+export const PrivateQuiz: React.FC = () => {
+    // @ts-ignore
+    const { key } = useParams();
+
+    const [isCorrectKey, setIsCorrectKey] = useState<boolean>(true);
+    const [showResult, setShowResult] = useState(false);
+    const [selected, setSelected] = useState<string | undefined>(undefined);
+
+    const [quiz, setQuiz] = useState<QuestionsSetResponse | undefined>(undefined);
+    const [numberOfQuestions, setNumberOfQuestions] = useState(0);
+    const [question, setQuestion] = useState<TeacherQuestionResponse | undefined>(undefined);
+    const [questionNumber, setQuestionNumber] = useState(1);
+
+    const [correctAnswers, setCorrectAnswers] = useState(0);
+    const [showLoading, setShowLoading] = useState(false);
+
+    useEffect(() => {
+        setShowLoading(true);
+        getQuiz().finally(() => setShowLoading(false));
+    }, []);
+
+    const getQuiz = () => {
+        return quizService.get(key)
+            .then(res => {
+                console.log(`Chosen key: ${key}`);
+                console.log(`Downloaded quiz: ${res.questionsSetName}`);
+
+                setQuiz(res);
+                setNumberOfQuestions(res.teacherQuestionsResponse.length);
+                setQuestion(res.teacherQuestionsResponse[0]);
+            })
+            .catch(err => {
+                console.log(err);
+                setIsCorrectKey(false);
+            });
+    }
+
+    const getNextQuestion = () => {
+        setShowResult(false);
+        setQuestionNumber(prev => prev+1);
+        setQuestion(quiz?.teacherQuestionsResponse[questionNumber])
```

Comment:
What happens here is misleading. You increment `questionNumber` and then use `setQuestion`, passsing to it previous `questionNumber` (before it is updated). It happens, because of the reason explained [here](https://dev.to/shareef/react-usestate-hook-is-asynchronous-1hia). It works correctly, because you manually set first `question` as that on index `0` in line 95. And then, when you call `getNextQuestion `for the first time, you increment `questionNumber` to `2`, and set `question` as that on index `1` (previous value of `questionNumber`).

As I said before, it works, so I think you don't have to fix it now (we're very short on time), but in the future I would use the solution from [here](https://dev.to/shareef/react-usestate-hook-is-asynchronous-1hia) - tl:dr `setQuesion` in `useEffect` with dependency `[questionNumber].`
Then, the initial value of `questionNumber` should be `0`. You can display `questionNumber+1` in UI.

Reply:
Done